airflow/security/tokens.py

```diff
@@ -0,0 +1,504 @@
+
+    return factory
+
+
+def _conf_list_factory(section, key, first_only: bool = False, **kwargs):
+    def factory() -> list[str] | str:
+        from airflow.configuration import conf
+
+        val = conf.getlist(section, key, **kwargs, suppress_warnings=True)
+
+        if first_only and val:
+            return val[0]
+        return val
+
+    return factory
+
+
+def _sec_to_timedelta(what: int | timedelta) -> timedelta:
+    if isinstance(what, timedelta):
+        return what
+    return timedelta(seconds=what)
+
+
+def _to_list(val: str | list[str]) -> list[str]:
+    if isinstance(val, str):
+        val = [val]
+    return val
+
+
+@attrs.define(kw_only=True)
+class JWTValidator:
+    """
+    Validate the claims and validitory of a JWT.
+
+    This will either validate the JWT is signed with the symmetric key if ``secret_key`` is passed, or else
+    that it is signed by one of the public keys in the keyset in ``jwks`` attribute.
+    """
+
+    jwks: JWKS | None = None
+    secret_key: str | None = attrs.field(repr=False, default=None, converter=lambda v: None if v == "" else v)
+    issuer: str | list[str] | None = attrs.field(
+        factory=_conf_list_factory("api_auth", "jwt_issuer", fallback=None)
+    )
+    # By default, we just validate these
+    required_claims: frozenset[str] = frozenset({"exp", "iat", "nbf"})
+    audience: str | Sequence[str]
+    algorithm: list[str] = attrs.field(default=["GUESS"], converter=_to_list)
+
+    leeway: timedelta = attrs.field(default=timedelta(seconds=5), converter=_sec_to_timedelta)
+
+    def __attrs_post_init__(self):
+        if not (self.jwks is None) ^ (self.secret_key is None):
+            raise ValueError("Exactly one of priavte_key and secret_key must be specified")
+
+        if self.algorithm == ["GUESS"]:
+            if self.jwks:
+                # TODO: We could probably populate this from the jwks document?
+                raise ValueError("Cannot guess the algorithm when using JWKS")
+            else:
+                self.algorithm = ["HS512"]
+
+    def _get_kid_from_header(self, unvalidated: str) -> str:
+        header = jwt.get_unverified_header(unvalidated)
+        if "kid" not in header:
+            raise ValueError("Missing 'kid' in token header")
+        return header["kid"]
+
+    async def _get_validation_key(self, unvalidated: str) -> str | jwt.PyJWK:
+        if self.secret_key:
+            return self.secret_key
+
+        if TYPE_CHECKING:
+            assert self.jwks is not None
+
+        kid = self._get_kid_from_header(unvalidated)
+        return await self.jwks.get_key(kid)
+
+    def validated_claims(
+        self, unvalidated: str, extra_claims: dict[str, Any] | None = None
+    ) -> dict[str, Any]:
+        return async_to_sync(self.avalidated_claims)(unvalidated, extra_claims)
+
+    async def avalidated_claims(
+        self, unvalidated: str, extra_claims: dict[str, Any] | None = None
```

Comment:
I'm not sure I understand the pattern for `extra_claims`. Claims are already contained in the token and returned when decoded, why do we need to validate also against extra claims ?

Reply:
YEah this looks like a mistake

Reply:
Ohh sorry, these are extra claims to validate. I'll rename the arg

i.e.

```python
await j.avalidated_claims(token, { "my_claim":{"essential": True, "value": "XYZ"}})
```